modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,479 @@
+            this::reducePreparationAndMetasCheck);
+
+        phase2PartsHashes = new DistributedProcess<>(kctx, SNAPSHOT_VALIDATE_PARTS, this::validateParts,
+            this::reduceValidatePartsAndFinish);
+
+        kctx.event().addLocalEventListener(evt -> onNodeLeft(((DiscoveryEvent)evt).eventNode().id()), EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /** Expected to run in a discovery-managed thread. */
+    private void onNodeLeft(UUID nodeId) {
+        Throwable err = new ClusterTopologyCheckedException("Snapshot checking stopped. " +
+            "A required node or the initiator node left the cluster [nodeId=" + nodeId + ']');
+
+        Iterator<Map.Entry<String, SnapshotCheckContext>> it = contexts.entrySet().iterator();
+
+        while (it.hasNext()) {
+            SnapshotCheckContext ctx = it.next().getValue();
+
+            if (!ctx.req.nodes().contains(nodeId))
+                continue;
+
+            if (ctx.fut != null)
+                ctx.fut.onDone(err);
+
+            it.remove();
+
+            GridFutureAdapter<?> fut = clusterOpFuts.get(ctx.req.reqId);
+
+            if (fut != null)
+                fut.onDone(err);
+        }
+    }
+
+    /** */
+    Map<String, SnapshotCheckContext> contexts() {
+        return Collections.unmodifiableMap(contexts);
+    }
+
+    /**
+     * Stops all the processes with the passed exception.
+     *
+     * @param err The interrupt reason.
+     */
+    void interrupt(Throwable err) {
+        contexts.forEach((snpName, ctx) -> {
+            if (ctx.fut != null)
+                ctx.fut.onDone(err);
+        });
+
+        clusterOpFuts.forEach((reqId, fut) -> fut.onDone(err));
+    }
+
+    /** Phase 2 and process finish. */
+    private IgniteInternalFuture<?> reduceValidatePartsAndFinish(
+        UUID reqId,
+        Map<UUID, SnapshotCheckResponse> results,
+        Map<UUID, Throwable> errors
+    ) {
+        SnapshotCheckContext ctx = context(null, reqId);
+
+        if (ctx != null) {
+            contexts.remove(ctx.req.snapshotName());
+
+            if (log.isInfoEnabled())
+                log.info("Finished snapshot local validation [req=" + ctx.req + ']');
+
+            GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(reqId);
+
+            // Operation node is the initiator collecting the final check operation result.
+            if (clusterOpFut != null) {
+                Map<ClusterNode, Exception> errors0 = mapErrors(errors);
+
+                if (!F.isEmpty(results)) {
+                    assert results.values().stream().noneMatch(res -> res != null && res.metas != null);
+
+                    Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> results0 = mapPartsHashes(results, ctx.req.nodes());
+
+                    IdleVerifyResultV2 chkRes = SnapshotChecker.reduceHashesResults(results0, errors0);
```

Comment:
It might throw error and then `clusterOpFut` will not finish.

Reply:
It doesn't declare an exception and doesn't throw it anywhere inside.